### PR TITLE
Add WebUI mock server and improve hyper-parameter parallel coordinates chart

### PR DESCRIPTION
**Pull request by ultmaster**

### CODE REVIEW

File: docs/en_US/Tutorial/SetupNniDeveloperEnvironment.md [hunk 1]
```diff
@@ -48,7 +48,8 @@ Nothing to do, the code is already linked to package folders.
 #### TypeScript
 
 * If `src/nni_manager` is changed, run `yarn watch` under this folder. It will watch and build code continually. The `nnictl` need to be restarted to reload NNI manager.
-* If `src/webui` or `src/nasui` are changed, run `yarn start` under the corresponding folder. The web UI will refresh automatically if code is changed.
+* If `src/webui` is changed, run `yarn dev`, which will run a mock API server and a webpack dev server simultaneously. Use `EXPERIMENT` environment variable (e.g., `mnist-tfv1-running`) to change mock data. Built-in mock experiments are listed in `src/webui/mock`.
```
Comment (colorjam) [inline docs/en_US/Tutorial/SetupNniDeveloperEnvironment.md:51]: better to provide the entire command such as `EXPERIMENT=mnist-tfv1-running yarn dev`

File: src/webui/src/static/model/trial.ts [hunk 2]
```diff
@@ -156,6 +191,45 @@ class Trial implements TableObj {
...
+        }
+    }
+
+    public metrics(space: MultipleAxes): Map<SingleAxis, any> {
+        const ret = new Map<SingleAxis, any>();
+        const unexpectedEntries = new Map<string, any>();
+        if (this.acc === undefined) {
+            return ret;
+        }
+        const acc = typeof this.acc === 'number' ? { default: this.acc } : this.acc;
```
Comment (Lijiaoa) [inline src/webui/src/static/model/trial.ts:217]: `accDictionary` in source  is === your `acc`

File: src/webui/src/static/model/searchspace.ts [hunk 3]
```diff
@@ -0,0 +1,185 @@
...
+    name = '';
+    fullName = '';
+
+    constructor(trials: TableObj[]) {
+        const columns = new Map<string, any[]>();
+        for (const trial of trials) {
+            if (trial.acc === undefined)
+                continue;
+            // TODO: handle more than number and object
+            const acc = typeof trial.acc === 'number' ? { default: trial.acc } : trial.acc;
```
Comment (Lijiaoa) [inline src/webui/src/static/model/searchspace.ts:166]: you can use this field `accDictionary ` in data source

File: src/webui/src/static/model/searchspace.ts [hunk 4]
```diff
@@ -0,0 +1,185 @@
...
+export class MetricSpace implements MultipleAxes {
+    axes = new Map<string, SingleAxis>();
+    name = '';
+    fullName = '';
+
+    constructor(trials: TableObj[]) {
+        const columns = new Map<string, any[]>();
+        for (const trial of trials) {
+            if (trial.acc === undefined)
+                continue;
```
Comment (Lijiaoa) [inline src/webui/src/static/model/searchspace.ts:164]: add {} , please

File: src/webui/src/static/model/searchspace.ts [hunk 5]
```diff
@@ -0,0 +1,185 @@
...
+                newSearchSpace.axes.set(key, new SimpleOrdinalAxis(key, key, 'choice', new Set(value).values()));
+            }
+        });
+        return newSearchSpace;
+    }
+}
+
+export class MetricSpace implements MultipleAxes {
+    axes = new Map<string, SingleAxis>();
+    name = '';
```
Comment (Lijiaoa) [inline src/webui/src/static/model/searchspace.ts:157]: also name question

File: src/webui/src/static/model/searchspace.ts [hunk 6]
```diff
@@ -0,0 +1,185 @@
...
+    'randint',
+    'uniform',
+    'quniform',
+    'loguniform',
+    'qloguniform',
+    'normal',
+    'qnormal',
+    'lognormal',
+    'qlognormal'
+];
```
Comment (Lijiaoa) [inline src/webui/src/static/model/searchspace.ts:17]: could you put SUPPORTED_SEARCH_SPACE_TYPE to const.ts

File: src/webui/src/static/model/searchspace.ts [hunk 7]
```diff
@@ -0,0 +1,185 @@
...
+    min: number = 0;
+    max: number = 0;
+    type: string;
+    name: string;
+    fullName: string;
+    scale: 'log' | 'linear';
+    nested = false;
+
+    constructor(name: string, fullName: string, type: string, value: any) {
+        this.name = name;
```
Comment (Lijiaoa) [inline src/webui/src/static/model/searchspace.ts:34]: I think name is a sensitive name.

File: src/webui/src/static/model/searchspace.ts [hunk 8]
```diff
@@ -0,0 +1,185 @@
...
+export class SearchSpace implements MultipleAxes {
+    axes = new Map<string, SingleAxis>();
+    name: string;
+    fullName: string;
+
+    constructor(name: string, fullName: string, searchSpaceSpec: any) {
+        this.name = name;
+        this.fullName = fullName;
+        if (searchSpaceSpec === undefined)
+            return;
```
Comment (Lijiaoa) [inline src/webui/src/static/model/searchspace.ts:102]: {}

File: src/webui/src/static/model/searchspace.ts [hunk 9]
```diff
@@ -0,0 +1,185 @@
...
+        Object.entries(searchSpaceSpec).forEach((item) => {
+            const key = item[0], spec = item[1] as any;
+            if (key === '_name') {
+                return;
+            } else if (spec._type === 'choice' || spec._type === 'layer_choice' || spec._type === 'input_choice') {
```
Comment (Lijiaoa) [inline src/webui/src/static/model/searchspace.ts:107]: ['x', 'xx', 'xx'].includes(spec._type)

File: src/webui/src/static/model/searchspace.ts [hunk 10]
```diff
@@ -0,0 +1,185 @@
...
+        this.scale = type.indexOf('log') !== -1 ? 'log' : 'linear';
+        if (type === 'randint') {
+            this.min = value[0];
+            this.max = value[1] - 1;
+        } else if (type.indexOf('uniform') !== -1) {
+            this.min = value[0];
+            this.max = value[1];
+        } else if (type.indexOf('normal') !== -1) {
+            const [mu, sigma] = [value[0], value[1]];
+            this.min = mu - 4 * sigma;
```
Comment (Lijiaoa) [inline src/webui/src/static/model/searchspace.ts:46]: why x 4

File: src/webui/src/static/model/searchspace.ts [hunk 11]
```diff
@@ -0,0 +1,185 @@
...
+    }
+}
+
+export class SearchSpace implements MultipleAxes {
+    axes = new Map<string, SingleAxis>();
+    name: string;
+    fullName: string;
+
+    constructor(name: string, fullName: string, searchSpaceSpec: any) {
+        this.name = name;
```
Comment (Lijiaoa) [inline src/webui/src/static/model/searchspace.ts:99]: also name question

File: src/webui/src/static/model/searchspace.ts (hunk 5 above)
Comment (ultmaster) [reply src/webui/src/static/model/searchspace.ts:157]: Changed to `baseName`

File: src/webui/src/static/model/searchspace.ts (hunk 9 above)
Comment (ultmaster) [reply src/webui/src/static/model/searchspace.ts:107]: Good suggestion!

File: src/webui/src/static/model/searchspace.ts (hunk 10 above)
Comment (ultmaster) [reply src/webui/src/static/model/searchspace.ts:46]: In normal distribution, [mu - 4*sigma, mu + 4*sigma] is a very safe domain of the values.

**Pull request merged**